File: components/password_manager/core/common/experiments.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/password_manager/core/common/experiments.h"

#include "base/command_line.h"
#include "base/metrics/field_trial.h"
#include "components/password_manager/core/common/password_manager_switches.h"

namespace password_manager {

bool ManageAccountLinkExperimentEnabled() {
  std::string group_name =
      base::FieldTrialList::FindFullName("PasswordLinkInSettings");

  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
  if (command_line->HasSwitch(switches::kDisablePasswordLink))
    return false;

  if (command_line->HasSwitch(switches::kEnablePasswordLink))
    return true;

  // To match Finch enabling the experiment by default, this method returns true
  // unless explicitly told the experiment is disabled. This ensures trybot
  // coverage of the enabled case.
  return group_name != "Disabled";
}

bool ForceSavingExperimentEnabled() {
  return base::CommandLine::ForCurrentProcess()->HasSwitch(
      password_manager::switches::kEnablePasswordForceSaving);
}

}  // namespace password_manager

File: remoting/webapp/base/js/experiments.js
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview
 * Class for enabling experimental features.
 */

'use strict';

/** @suppress {duplicate} */
var remoting = remoting || {};

(function () {

var kExperimentsStorageName = 'remoting-experiments';

/**
 * @param {Array.<string>} list
 */
function save(list) {
  var storageMap = {};
  storageMap[kExperimentsStorageName] = list;
  chrome.storage.local.set(storageMap);
};

/** @type {Object} */
remoting.experiments = {};

/**
 * Enables an experiment.
 *
 * @param {string} experiment to enable.
 */
remoting.experiments.enable = function(experiment) {
  remoting.experiments.get().then(function(/** Array.<string> */list) {
    if (list.indexOf(experiment) == -1) {
      list.push(experiment);
      save(list);
    }
  });
};

/**
 * Disables an experiment.
 *
 * @param {string} experiment to disable.
 */
remoting.experiments.disable = function(experiment) {
  remoting.experiments.get().then(function(/** Array.<string> */list) {
    list = list.filter(function(e) { return e !== experiment; });
    save(list);
  });
};

/**
 * Returns list of all enabled experiments.
 * @return {Promise}
 */
remoting.experiments.get = function() {
  return new Promise(function(resolve, reject) {
    chrome.storage.local.get(kExperimentsStorageName, function(items) {
      if (items.hasOwnProperty(kExperimentsStorageName)) {
        resolve(items[kExperimentsStorageName]);
      }
      resolve([]);
    });
  });
};

})();
